File: scripts/fetch_rss.py
#!/usr/bin/env python3
"""
RSS取得スクリプト - GitHub Actions用

省庁のRSSフィードを取得し、JSONファイルとして保存します。
既存データとマージして履歴を保持します。
"""

import json
import urllib.request
import re
import html
from datetime import datetime, timezone
from pathlib import Path
from email.utils import parsedate_to_datetime
import hashlib

# プロジェクトルート
PROJECT_ROOT = Path(__file__).parent.parent

# 設定ファイルを読み込み
MINISTRIES_PATH = PROJECT_ROOT / "src/config/ministries.json"
OUTPUT_DIR = PROJECT_ROOT / "public/data"
OUTPUT_FILE = OUTPUT_DIR / "items.json"


def fetch_url(url: str) -> str | None:
    """URLのコンテンツを取得する"""
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req, timeout=15) as response:
            raw_bytes = response.read()
            for encoding in ['utf-8', 'shift_jis', 'euc-jp', 'iso-2022-jp']:
                try:
                    return raw_bytes.decode(encoding)
                except Exception:
                    continue
            return raw_bytes.decode('utf-8', errors='ignore')
    except Exception as e:
        print(f"  Error fetching {url}: {e}")
        return None


def discover_feed_urls(html_text: str, base_url: str) -> list[str]:
    """HTMLページからRSS/Atomフィードのリンクを探す"""
    urls = []
    seen = set()

    # <link rel="alternate" type="application/rss+xml" ...>
    link_re = re.compile(r'<link[^>]+>', re.IGNORECASE)
    for tag in link_re.findall(html_text):
        if not re.search(r'rel\s*=\s*["\']alternate["\']', tag, re.IGNORECASE):
            continue
        if not re.search(
            r'type\s*=\s*["\']application/(rss|atom)\+xml["\']', tag, re.IGNORECASE
        ):
            continue
        m = re.search(r'href\s*=\s*["\']([^"\']+)["\']', tag, re.IGNORECASE)
        if m:
            url = normalize_url(m.group(1), base_url)
            if url and url not in seen:
                seen.add(url)
                urls.append(url)

    # <a href="...rdf|...xml|...atom">
    anchor_re = re.compile(r'<a[^>]+href\s*=\s*["\']([^"\']+)["\'][^>]*>', re.IGNORECASE)
    for m in anchor_re.finditer(html_text):
        href = m.group(1)
        if re.search(r'\.(rdf|xml|atom)(\?|$)', href, re.IGNORECASE):
            url = normalize_url(href, base_url)
            if url and url not in seen:
                seen.add(url)
                urls.append(url)

    return urls


def normalize_url(url: str, base_url: str) -> str | None:
    """相対URLを絶対URLに変換する"""
    url = html.unescape(url).strip()
    if not url:
        return None
    if url.startswith('http://') or url.startswith('https://'):
        return url
    # 相対URL処理
    if url.startswith('//'):
        scheme = base_url.split('://')[0] if '://' in base_url else 'https'
        return f"{scheme}:{url}"
    if url.startswith('/'):
        m = re.match(r'(https?://[^/]+)', base_url)
        if m:
            return m.group(1) + url
        return None
    # 相対パス
    base = re.sub(r'/[^/]*$', '/', base_url)
    return base + url


def fetch_rss(url: str) -> list[dict]:
    """RSSフィードを取得してパースする"""
    text = fetch_url(url)
    if text is None:
        return []

    items = []

    # RSS形式
    for match in re.finditer(r'<item[^>]*>(.*?)</item>', text, re.DOTALL):
        item = parse_item(match.group(1), 'rss')
        if item:
            items.append(item)

    # Atom形式
    if not items:
        for match in re.finditer(r'<entry[^>]*>(.*?)</entry>', text, re.DOTALL):
            item = parse_item(match.group(1), 'atom')
            if item:
                items.append(item)

    return items


def fetch_rss_discovery(discovery_url: str, site_url: str) -> list[dict]:
    """HTMLページからRSSフィードURLを発見して取得する"""
    print(f"  Discovering feeds from: {discovery_url}")
    html_text = fetch_url(discovery_url)
    if html_text is None:
        return []

    feed_urls = discover_feed_urls(html_text, site_url)
    print(f"  Found {len(feed_urls)} feed(s): {feed_urls}")

    all_items = []
    for feed_url in feed_urls:
        items = fetch_rss(feed_url)
        print(f"    [{feed_url}] {len(items)} items")
        all_items.extend(items)

    return all_items


def parse_item(item_xml: str, feed_type: str) -> dict | None:
    """アイテムXMLをパースする"""
    try:
        # タイトル
        title_match = re.search(r'<title[^>]*>(.*?)</title>', item_xml, re.DOTALL)
        if not title_match:
            return None
        title = clean_text(title_match.group(1))
        if not title:
            return None

        # リンク
        if feed_type == 'atom':
            link_match = re.search(r'<link[^>]*href=["\']([^"\']+)["\']', item_xml)
        else:
            link_match = re.search(r'<link[^>]*>(.*?)</link>', item_xml, re.DOTALL)

        if not link_match:
            return None
        link = clean_text(link_match.group(1))
        if not link:
            return None

        # 日付
        date_patterns = [
            r'<pubDate[^>]*>(.*?)</pubDate>',
            r'<dc:date[^>]*>(.*?)</dc:date>',
            r'<published[^>]*>(.*?)</published>',
            r'<updated[^>]*>(.*?)</updated>',
        ]

        published_at = None
        date_estimated = True

        for pattern in date_patterns:
            date_match = re.search(pattern, item_xml, re.DOTALL)
            if date_match:
                date_str = date_match.group(1).strip()
                try:
                    published_at = parsedate_to_datetime(date_str).isoformat()
                    date_estimated = False
                    break
                except Exception:
                    try:
                        dt = datetime.fromisoformat(date_str.replace('Z', '+00:00'))
                        published_at = dt.isoformat()
                        date_estimated = False
                        break
                    except Exception:
                        continue

        # 日付が取得できない場合は現在時刻
        if not published_at:
            published_at = datetime.now(timezone.utc).isoformat()
            date_estimated = True

        return {
            'title': title,
            'url': link,
            'published_at': published_at,
            'date_estimated': date_estimated,
        }

    except Exception as e:
        print(f"  Error parsing item: {e}")
        return None


def clean_text(text: str) -> str:
    """テキストをクリーンアップ"""
    text = re.sub(r'<!\[CDATA\[(.*?)\]\]>', r'\1', text, flags=re.DOTALL)
    text = html.unescape(text)
    text = text.strip()
    return text


def make_item_hash(title: str, url: str) -> str:
    return hashlib.md5(f"{title}{url}".encode('utf-8')).hexdigest()


def load_existing_items() -> dict[str, dict]:
    """既存のitems.jsonを読み込む（id → item の辞書）"""
    if not OUTPUT_FILE.exists():
        return {}
    try:
        with open(OUTPUT_FILE, 'r', encoding='utf-8') as f:
            items = json.load(f)
        return {item['id']: item for item in items if 'id' in item}
    except Exception as e:
        print(f"Warning: could not load existing items: {e}")
        return {}


def main():
    """メイン処理"""
    print("=" * 60)
    print("RSS Feed Fetcher")
    print("=" * 60)

    # 設定ファイルを読み込み
    with open(MINISTRIES_PATH, 'r', encoding='utf-8') as f:
        ministries = json.load(f)

    print(f"\nLoaded {len(ministries)} ministries")

    # 既存データをロード（マージ用）
    existing_items = load_existing_items()
    print(f"Existing items: {len(existing_items)}")

    # 新規フェッチしたアイテムを格納
    new_items: dict[str, dict] = {}
    fetched_at = datetime.now(timezone.utc).isoformat()

    # 各省庁のRSSを取得
    for ministry in ministries:
        ministry_name = ministry['ministry']
        site_url = ministry.get('siteUrl', '')
        print(f"\n{ministry_name}")
        print("-" * 40)

        for source in ministry['sources']:
            source_name = source['name']
            source_url = source['url']
            source_type = source.get('type', 'rss')

            print(f"  [{source_name}] Fetching... ({source_type})")

            if source_type == 'rss-discovery':
                items = fetch_rss_discovery(source_url, site_url)
            else:
                items = fetch_rss(source_url)

            print(f"  [{source_name}] Found {len(items)} items")

            for item in items:
                item['url'] = html.unescape(item['url']).strip()
                item_hash = make_item_hash(item['title'], item['url'])

                if item_hash in new_items:
                    continue

                item['id'] = item_hash
                item['ministry'] = ministry_name
                item['source_name'] = source_name
                item['fetched_at'] = fetched_at
                new_items[item_hash] = item

    # 既存データとマージ（既存を保持しつつ新規で上書き）
    merged: dict[str, dict] = {**existing_items, **new_items}

    all_items = sorted(
        merged.values(),
        key=lambda x: x.get('published_at', ''),
        reverse=True,
    )

    # JSONファイルに保存
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(all_items, f, ensure_ascii=False, indent=2)

    print("\n" + "=" * 60)
    print(f"New items fetched : {len(new_items)}")
    print(f"Total items saved : {len(all_items)}")
    print(f"Output: {OUTPUT_FILE}")
    print("=" * 60)

    # サマリー
    print("\nSummary by ministry:")
    ministry_counts: dict[str, int] = {}
    for item in all_items:
        m = item.get('ministry', 'unknown')
        ministry_counts[m] = ministry_counts.get(m, 0) + 1
    for m, count in sorted(ministry_counts.items()):
        print(f"  {m}: {count} items")


if __name__ == '__main__':
    main()

File: scripts/fetch_rss.mjs
#!/usr/bin/env node
/**
 * RSS取得スクリプト (Node.js版) - Vercelビルド / GitHub Actions用
 *
 * 各省庁のRSSフィードを取得し public/data/items.json へマージ保存する。
 * Python不要。Node.js >=18 の fetch API を使用。
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from "fs";
import { createHash } from "crypto";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const PROJECT_ROOT = join(__dirname, "..");

const MINISTRIES_PATH = join(PROJECT_ROOT, "src/config/ministries.json");
const OUTPUT_DIR = join(PROJECT_ROOT, "public/data");
const OUTPUT_FILE = join(OUTPUT_DIR, "items.json");

/* ─── helpers ─── */

function md5(str) {
  return createHash("md5").update(str, "utf-8").digest("hex");
}

function decodeEntities(s) {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'");
}

function cleanText(raw) {
  let t = raw.replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, "$1");
  t = decodeEntities(t);
  t = t.replace(/<[^>]*>/g, "");
  return t.trim();
}

function resolveUrl(href, base) {
  try {
    return new URL(decodeEntities(href.trim()), base).toString();
  } catch {
    return null;
  }
}

/** 日付文字列 → ISO string。パース不可なら null */
function parseDate(raw) {
  if (!raw) return null;
  const s = raw.trim();
  const d = new Date(s);
  if (!isNaN(d.getTime())) return d.toISOString();
  // ISO 8601 の +0900 形式
  const d2 = new Date(s.replace(/([+-]\d{2})(\d{2})$/, "$1:$2"));
  if (!isNaN(d2.getTime())) return d2.toISOString();
  return null;
}

/* ─── fetch with timeout & retry ─── */

async function fetchText(url, forcedEncoding = null, retries = 2) {
  for (let i = 0; i <= retries; i++) {
    try {
      const res = await fetch(url, {
        headers: {
          "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
          "Accept": "application/rss+xml, application/atom+xml, application/xml, text/xml, */*",
        },
        signal: AbortSignal.timeout(15_000),
        redirect: "follow",
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const buf = await res.arrayBuffer();

      // 1. Use forced encoding from source config if provided
      if (forcedEncoding) {
        try {
          return new TextDecoder(forcedEncoding).decode(buf);
        } catch {
          // fall through to auto-detection
        }
      }

      // 2. Check HTTP Content-Type header for charset
      const ct = res.headers.get("content-type") || "";
      const ctMatch = ct.match(/charset\s*=\s*([^\s;]+)/i);
      if (ctMatch) {
        const ctEnc = ctMatch[1].toLowerCase().replace(/^["']|["']$/g, "");
        if (ctEnc !== "utf-8" && ctEnc !== "utf8") {
          try {
            return new TextDecoder(ctEnc).decode(buf);
          } catch { /* fall through */ }
        }
      }

      // 3. Detect encoding from XML declaration (<?xml ... encoding="Shift_JIS"?>)
      const head = new TextDecoder("latin1").decode(new Uint8Array(buf).slice(0, 512));
      const encMatch = head.match(/encoding\s*=\s*["']([^"']+)["']/i);
      if (encMatch) {
        const declaredEnc = encMatch[1].toLowerCase();
        if (declaredEnc !== "utf-8" && declaredEnc !== "utf8") {
          try {
            return new TextDecoder(declaredEnc).decode(buf);
          } catch { /* fall through */ }
        }
      }

      // 4. Default: UTF-8
      return new TextDecoder("utf-8").decode(buf);
    } catch (e) {
      if (i === retries) {
        console.error(`  FAIL ${url}: ${e.message}`);
        return null;
      }
      await new Promise((r) => setTimeout(r, 1000 * (i + 1)));
    }
  }
  return null;
}

/* ─── RSS/Atom parsing ─── */

function parseItems(xml) {
  const items = [];

  // RSS 1.0 (RDF) の <channel> と <image> ブロックを除去してから解析
  // （チャンネル画像が <item> と誤認識されるのを防ぐ）
  const stripped = xml
    .replace(/<channel[\s>][\s\S]*?<\/channel>/gi, "")
    .replace(/<image[\s>][\s\S]*?<\/image>/gi, "");

  // RSS <item>
  for (const m of stripped.matchAll(/<item[^>]*>([\s\S]*?)<\/item>/gi)) {
    const it = parseEntry(m[1], "rss");
    if (it) items.push(it);
  }

  // Atom <entry>
  if (items.length === 0) {
    for (const m of stripped.matchAll(/<entry[^>]*>([\s\S]*?)<\/entry>/gi)) {
      const it = parseEntry(m[1], "atom");
      if (it) items.push(it);
    }
  }

  // RDF <item> (older format)
  if (items.length === 0) {
    for (const m of stripped.matchAll(/<item\s[^>]*>([\s\S]*?)<\/item>/gi)) {
      const it = parseEntry(m[1], "rss");
      if (it) items.push(it);
    }
  }

  return items;
}

function parseEntry(block, type) {
  const titleM = block.match(/<title[^>]*>([\s\S]*?)<\/title>/i);
  if (!titleM) return null;
  const title = cleanText(titleM[1]);
  if (!title) return null;

  let link = null;
  if (type === "atom") {
    const m = block.match(/<link[^>]*href=["']([^"']+)["'][^>]*\/?>/i);
    if (m) link = decodeEntities(m[1].trim());
  } else {
    const m = block.match(/<link[^>]*>([\s\S]*?)<\/link>/i);
    if (m) link = cleanText(m[1]);
  }
  if (!link) return null;

  // date
  const dateCandidates = [
    /<pubDate[^>]*>([\s\S]*?)<\/pubDate>/i,
    /<dc:date[^>]*>([\s\S]*?)<\/dc:date>/i,
    /<published[^>]*>([\s\S]*?)<\/published>/i,
    /<updated[^>]*>([\s\S]*?)<\/updated>/i,
  ];
  let published_at = null;
  let date_estimated = true;
  for (const re of dateCandidates) {
    const dm = block.match(re);
    if (dm) {
      const parsed = parseDate(dm[1].trim());
      if (parsed) {
        published_at = parsed;
        date_estimated = false;
        break;
      }
    }
  }
  if (!published_at) {
    published_at = new Date().toISOString();
    date_estimated = true;
  }

  return { title, url: link, published_at, date_estimated };
}

/* ─── RSS discovery (HTML → feed links) ─── */

function discoverFeedUrls(html, baseUrl) {
  const urls = new Set();

  // <link rel="alternate" type="application/rss+xml" href="...">
  for (const m of html.matchAll(/<link[^>]+>/gi)) {
    const tag = m[0];
    if (!/rel\s*=\s*["']alternate["']/i.test(tag)) continue;
    if (
      !/type\s*=\s*["']application\/(rss|atom)\+xml["']/i.test(tag) &&
      !/type\s*=\s*["']text\/xml["']/i.test(tag)
    )
      continue;
    const hrefM = tag.match(/href\s*=\s*["']([^"']+)["']/i);
    if (hrefM) {
      const resolved = resolveUrl(hrefM[1], baseUrl);
      if (resolved) urls.add(resolved);
    }
  }

  // <a href="...rdf|xml|atom">
  for (const m of html.matchAll(/<a[^>]+href\s*=\s*["']([^"']+)["'][^>]*>/gi)) {
    const href = m[1];
    if (/\.(rdf|xml|atom)(\?|#|$)/i.test(href)) {
      const resolved = resolveUrl(href, baseUrl);
      if (resolved) urls.add(resolved);
    }
  }

  return [...urls];
}

/* ─── main ─── */

async function main() {
  console.log("=".repeat(60));
  console.log("RSS Feed Fetcher (Node.js)");
  console.log("=".repeat(60));

  const ministries = JSON.parse(readFileSync(MINISTRIES_PATH, "utf-8"));
  console.log(`\nLoaded ${ministries.length} ministries`);

  // 既存データ読み込み
  let existing = {};
  if (existsSync(OUTPUT_FILE)) {
    try {
      const arr = JSON.parse(readFileSync(OUTPUT_FILE, "utf-8"));
      for (const item of arr) {
        if (item.id) existing[item.id] = item;
      }
    } catch {}
  }
  console.log(`Existing items: ${Object.keys(existing).length}`);

  const fetched = {};
  const now = new Date().toISOString();

  for (const ministry of ministries) {
    const name = ministry.ministry;
    const siteUrl = ministry.siteUrl || "";
    console.log(`\n${name}`);
    console.log("-".repeat(40));

    for (const source of ministry.sources) {
      const { name: srcName, url: srcUrl, type: srcType = "rss", encoding: srcEncoding = null } = source;
      console.log(`  [${srcName}] Fetching... (${srcType}${srcEncoding ? ", " + srcEncoding : ""})`);

      let items = [];

      if (srcType === "rss-discovery") {
        const html = await fetchText(srcUrl, srcEncoding);
        if (html) {
          const feedUrls = discoverFeedUrls(html, siteUrl);
          console.log(`  [${srcName}] Discovered ${feedUrls.length} feed(s)`);
          for (const feedUrl of feedUrls) {
            const xml = await fetchText(feedUrl, srcEncoding);
            if (xml) {
              const parsed = parseItems(xml);
              console.log(`    ${feedUrl} → ${parsed.length} items`);
              items.push(...parsed);
            }
          }
        }
      } else {
        const xml = await fetchText(srcUrl, srcEncoding);
        if (xml) items = parseItems(xml);
      }

      console.log(`  [${srcName}] Found ${items.length} items`);

      for (const item of items) {
        item.url = resolveUrl(item.url, siteUrl) || item.url;

        // ホームページ直リンク・画像URLはゴミアイテムとしてスキップ
        try {
          const u = new URL(item.url);
          if (u.pathname === "/" || u.pathname === "") continue;
          if (/\.(gif|jpg|jpeg|png|svg|webp|ico)(\?|#|$)/i.test(u.pathname)) continue;
        } catch { continue; }

        // URL のみで重複排除（タイトルが変化しても同一記事として扱い、再取得で上書き可能にする）
        const hash = md5(item.url);
        if (fetched[hash]) continue;
        fetched[hash] = {
          ...item,
          id: hash,
          ministry: name,
          source_name: srcName,
          fetched_at: now,
        };
      }
    }
  }

  // merge（90日以上前のアイテムは除外）
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - 90);
  const cutoffStr = cutoff.toISOString();

  const merged = { ...existing, ...fetched };

  // URL-based deduplication: keep the most recently fetched item per URL
  // This handles the case where old items have md5(title+url) IDs and new ones have md5(url) IDs
  const byUrl = new Map();
  for (const item of Object.values(merged)) {
    const url = item.url;
    const existing_for_url = byUrl.get(url);
    if (!existing_for_url || (item.fetched_at || "") >= (existing_for_url.fetched_at || "")) {
      byUrl.set(url, item);
    }
  }

  const all = Array.from(byUrl.values())
    .filter((item) => !item.date_estimated && (item.published_at || "") >= cutoffStr)
    .sort(
      (a, b) => (b.published_at || "").localeCompare(a.published_at || ""),
    );

  mkdirSync(OUTPUT_DIR, { recursive: true });
  writeFileSync(OUTPUT_FILE, JSON.stringify(all, null, 2), "utf-8");

  console.log("\n" + "=".repeat(60));
  console.log(`New items fetched : ${Object.keys(fetched).length}`);
  console.log(`Total items saved : ${all.length}`);
  console.log(`Output: ${OUTPUT_FILE}`);
  console.log("=".repeat(60));

  // summary
  const counts = {};
  for (const item of all) {
    const m = item.ministry || "unknown";
    counts[m] = (counts[m] || 0) + 1;
  }
  console.log("\nSummary by ministry:");
  for (const [m, c] of Object.entries(counts).sort()) {
    console.log(`  ${m}: ${c} items`);
  }
}

main().catch((e) => {
  console.error("Fatal:", e);
  // ビルドを失敗させない（既存のitems.jsonでフォールバック）
  process.exit(0);
});
